blog/zk-snark-pallet.md

```diff
@@ -0,0 +1,263 @@
+
+#### #[pallet::config]
+All pallets in substrates define a trait called `Config`, which needs to be defined under this macro. We can declare here some specific pallet requirements. In our case for the zk-SNARK we are going to define some constant values:
+```
+#[pallet::constant]
+type MaxPublicInputsLength: Get<u32>;
+
+#[pallet::constant]
+type MaxProofLength: Get<u32>;
+
+#[pallet::constant]
+type MaxVerificationKeyLength: Get<u32>;
+```
+As you probably guess, we are going to use those constants for checking the maximum length of our input files.
+
+#### #[pallet::event]
+This section defines all events that can be emitted from our pallet. We have defined several events here that left information on the blockchain about the progress of the verification. 
+
+```
+#[pallet::event]
+#[pallet::generate_deposit(pub(super) fn deposit_event)]
+pub enum Event<T: Config> {
+    VerificationSetupCompleted,
+    VerificationProofSet,
+    VerificationSuccess { who: T::AccountId },
+    VerificationFailed,
+}
+```
+What is worth mentioning, Rust language allows the definition of different types of enums fields. We take advantage of this feature when we have declared a `VerificationSuccess` event. The event will store an `AccountId`, which represents an account that belongs to the person who submits valid proof.
+
+#### #[pallet::error]
+Here we define all errors, which our pallet returns when something goes wrong.
+```
+#[pallet::error]
+pub enum Error<T> {
+    PublicInputsMismatch,
+    TooLongPublicInputs,
+    TooLongVerificationKey,
+    TooLongProof,
+    ProofIsEmpty,
+    VerificationKeyIsNotSet,
+    MalformedVerificationKey,
+    MalformedProof,
+    MalformedPublicInputs,
+    NotSupportedCurve,
+    NotSupportedProtocol,
+    ProofVerificationError,
+    ProofCreationError,
+    VerificationKeyCreationError,
+}
+```
+
+#### #[pallet::storage]
+The storage section defines all information that can be stored on the blockchain. In the context of zk-SNARKs, we would like to store artifacts generated by Bob (public inputs, verification key) and Alice (proof).
+```
+#[pallet::storage]
+pub type PublicInputStorage<T: Config> = StorageValue<_, PublicInputsDef<T>, ValueQuery>;
+
+#[pallet::storage]
+pub type ProofStorage<T: Config> = StorageValue<_, ProofDef<T>, ValueQuery>;
+
+#[pallet::storage]
+pub type VerificationKeyStorage<T: Config> = StorageValue<_, VerificationKeyDef<T>, ValueQuery>;
+
+```
+
+#### #[pallet::call]
+Final section declarations extrinsics, which is the interface for pallets. As we mentioned earlier, we defined two methods. Thanks to them, we will be able to interact with the zk-SNARK pallet.
+
+The first method is for the verification setup, which is going to be used by Bob to set up the contest. We are going to store public input and the verification key. We will emit a *VerificationSetupCompleted* event, and if anything goes wrong, we will return an appropriate error.
```

Comment:
Proponuję rozbić na osobne zdania: We will emit a *VerificationSetupCompleted* event. If anything goes wrong, we will return an appropriate error.